apm-agent-plugins/apm-grpc/apm-grpc-plugin/src/main/java/co/elastic/apm/agent/grpc/GrpcHelper.java

```diff
@@ -312,25 +323,83 @@ public Span startSpan(@Nullable AbstractSpan<?> parent,
     }
 
     /**
+     * Called at the exit from {@link io.grpc.Channel#newCall(MethodDescriptor, CallOptions)}.
      * Registers (and deactivates) span in internal storage for lookup by client call.
      * <br>
-     * This is the 2cnd method called during client call execution, the next is {@link #clientCallStartEnter(ClientCall, ClientCall.Listener, Metadata)}.
+     * This is the 2nd method called during client call execution, the next is {@link #clientCallStartEnter(ClientCall, ClientCall.Listener, Metadata)}.
      *
-     * @param clientCall client call
-     * @param span       span
+     * @param clientCall    client call
+     * @param spanFromEntry span created at {@link #onClientCallCreationEntry(AbstractSpan, MethodDescriptor, String)}
      */
-    public void registerSpan(@Nullable ClientCall<?, ?> clientCall, Span span) {
+    public void onClientCallCreationExit(@Nullable ClientCall<?, ?> clientCall, @Nullable Span spanFromEntry) {
         if (clientCall != null) {
-            clientCallSpans.put(clientCall, span);
+            Span spanToMap = spanFromEntry;
+            if (spanToMap == null) {
+                // handling nested newCall() invocations - we still want to map the client call to the same span
+                Span tmp = GlobalTracer.get().getActiveSpan();
+                if (tmp != null && tmp.getSubtype() != null && tmp.getSubtype().equals(GRPC) && tmp.isExit()) {
+                    spanToMap = tmp;
+                }
+            }
+
+            if (spanToMap != null && !spanToMap.isDiscarded()) {
```

Comment:
[question] in which case does the span would have been discarded ?

Reply:
Not sure if it's a real possibility, but since we are discarding spans I think it worth the check - we don't need to continue with mappings for discarded spans.
Theoretical option: two internal gRPC channels delegate `newCall() ` from one to the other, where the nested call already sets the created call as the real call in the corresponding `DelayedClientCall`, causing the active span to be discarded already before the outer call exits.